test/distributed/org/apache/cassandra/distributed/upgrade/MixedModePaxosTest.java

```diff
@@ -0,0 +1,226 @@
+     * ttl'd paxos data though, the ttl'd data on disk would shadow this update and the conflicting 'zero' value would
+     * cause the prepare phase to get into an infinite loop.
+     */
+    @Test
+    public void upgradeAwareTtldPaxosStateTest() throws Throwable
+    {
+        String keyspace = KEYSPACE;
+        String table = "tbl";
+        int gcGrace = 10; // 1 day
+        new TestCase()
+        .withConfig(c -> c.with(Feature.GOSSIP, Feature.NETWORK))
+        .nodes(2)
+        .nodesToUpgrade(1)
+        // all upgrades from v30 up, excluding v30->v3X and from v40
+        .singleUpgrade(v40)
+        .setup(cluster -> {
+            cluster.schemaChange(format("CREATE TABLE %s.%s (k int primary key, v int) " +
+                                        "WITH gc_grace_seconds=%s", keyspace, table, gcGrace));
+        })
+        .runAfterClusterUpgrade(cluster -> {
+            // disable compaction to prevent paxos state from being purged
+            cluster.forEach(instance -> instance.nodetool("disableautocompaction"));
+
+            // insert a ttl'd committed paxos state
+            long ballotMicros = TimeUnit.NANOSECONDS.toMicros(System.currentTimeMillis());
+            FakePaxosHelper helper = FakePaxosHelper.create(cluster.coordinator(1), keyspace, table, gcGrace, ballotMicros);
+
+            // confirm none of the nodes have paxos state
+            for (int i=1; i<=cluster.size(); i++)
+                Assert.assertEquals(0, cluster.coordinator(i).execute("SELECT * FROM system.paxos", ConsistencyLevel.ONE).length);
+
+
+            // save commit to both nodes
+            helper.saveCommit(cluster.coordinator(1));
+            helper.saveCommit(cluster.coordinator(2));
+
+            // wait for the paxos state to expire
+            Thread.sleep(TimeUnit.SECONDS.toMillis(gcGrace * 2));
+
+            // confirm paxos state on node 2 has ttld
+            Assert.assertEquals(0, cluster.coordinator(1).execute("SELECT * FROM system.paxos", ConsistencyLevel.ONE).length);
+            Assert.assertEquals(0, cluster.coordinator(2).execute("SELECT * FROM system.paxos", ConsistencyLevel.ONE).length);
+
+            // paxos operation should not timeout
+            cluster.coordinator(1).execute(format("SELECT * FROM %s.%s WHERE k=1", keyspace, table), ConsistencyLevel.SERIAL);
+
+
+        })
+        .run();
+    }
+
+    /**
+     * Variant of the above test where the upgraded node has not seen the ttl'd commit
+     * @throws Throwable
+     */
+    @Test
+    public void upgradeUnawareTtldPaxosStateTest() throws Throwable
+    {
+        String keyspace = KEYSPACE;
+        String table = "tbl";
+        int gcGrace = 10; // 1 day
+        new TestCase()
+        .withConfig(c -> c.with(Feature.GOSSIP, Feature.NETWORK))
+        .nodes(2)
+        .nodesToUpgrade(1)
+        // all upgrades from v30 up, excluding v30->v3X and from v40
```

Comment:
This also seems like maybe a C&P comment that doesn't apply since it mentions additional upgrades?